### PR TITLE
New API keys

### PR DESCRIPTION
**Pull request by AntonCroner**

### CODE REVIEW

File: Authenticator.spec.ts [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+	it("one realm role", async () => {
+		const realm: pax2pay.Realm | undefined = undefined
+		const token = await issuer.sign({
+			name: { first: "", last: "" },
+			email: "",
+			role: JSON.stringify({ test: "finance" }),
+			permissions: "",
+		})
+		const authenticated = await authenticator.authenticate(token ?? "", {
+			[realm ?? "*"]: { treasury: { rebalance: true } },
```
Comment (marcushanses) [inline Authenticator.spec.ts:21]: Realm is always undefined.

File: Authenticator.spec.ts [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+		})
+		console.log("authenticated: ", authenticated)
```
Comment (marcushanses) [inline Authenticator.spec.ts:23]: Remove log.

File: Authenticator.spec.ts [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+		expect(authenticated).toBeTruthy()
+	})
+	it("several realms role example", async () => {
+		const realm: pax2pay.Realm | undefined = "test"
+		const token = await issuer.sign({
```
Comment (marcushanses) [inline Authenticator.spec.ts:28]: Make a function instead of writing the same code over and over...

File: Authenticator.spec.ts [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+			name: { first: "", last: "" },
+			email: "",
+			role: JSON.stringify({ test: "finance", testUK: "finance" }),
+			permissions: "",
+		})
+		const authenticated = await authenticator.authenticate(token ?? "", {
```
Comment (marcushanses) [inline Authenticator.spec.ts:34]: This function should be able to handle undefined so we don't have to use '?? ""' everywhere.

File: Authenticator.spec.ts [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+	const authenticator = new pax2pay.Authenticator(publicKey)
+	it("one realm role", async () => {
+		const realm: pax2pay.Realm | undefined = undefined
+		const token = await issuer.sign({
+			name: { first: "", last: "" },
+			email: "",
+			role: JSON.stringify({ test: "finance" }),
+			permissions: "",
+		})
+		const authenticated = await authenticator.authenticate(token ?? "", {
```
Comment (marcushanses) [inline Authenticator.spec.ts:20]: Make the function accept undefined tokens.
.

File: Authenticator.spec.ts [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+	it("all realms role example", async () => {
+		const realm: pax2pay.Realm | undefined = "test"
+		const token = await issuer.sign({
+			name: { first: "", last: "" },
+			email: "",
+			role: JSON.stringify({ "*": "finance" }),
+			permissions: "",
+		})
+		const authenticated = await authenticator.authenticate(token ?? "", {
+			[realm ?? "*"]: { treasury: { rebalance: true } },
```
Comment (marcushanses) [inline Authenticator.spec.ts:52]: Realm is a string!

File: Authenticator.spec.ts [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+	it("permissions example one organization", async () => {
+		const organization: string | undefined = undefined
+		const permissions: pax2pay.Key.Permissions = { asdf: { cards: true } }
+		const token = await issuer.sign({
+			name: { first: "", last: "" },
+			email: "",
+			permissions: flagly.Flags.stringify(permissions),
+		})
+		const authenticated = await authenticator.authenticate(token ?? "", {
+			[organization ?? "*"]: { cards: { view: true } },
```
Comment (marcushanses) [inline Authenticator.spec.ts:69]: Organizaion is a const defined above.

File: Authenticator.spec.ts [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+		const authenticated = await authenticator.authenticate(token ?? "", {
+			[realm ?? "*"]: { cards: { view: true } },
+		})
+		const authenticatedNoRealm = await authenticator.authenticate(token ?? "", {
+			["*"]: { cards: { view: true } },
+		})
+		expect(authenticated).toBeTruthy()
+		expect(authenticatedNoRealm).toBeTruthy()
+	})
+})
```
Comment (marcushanses) [inline Authenticator.spec.ts:154]: This is just the same code over and over. Also most comments apply to several rows but I don't want to repeat myself.

File: Authenticator.ts [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+	},
+}
+
+export class Authenticator {
+	private readonly verifier: userwidgets.User.Key.Verifier<Key> = userwidgets.User.Key.Verifier.create<Key>(this.key)
+
+	constructor(private readonly key: string) {
+		this.verifier.add(roleTransformer)
+	}
+	async authenticate(token: string, permissions: Key.Permissions): Promise<Key | false | undefined> {
```
Comment (marcushanses) [inline Authenticator.ts:23]: Shouldn't this function accept roles or permissions?

File: Authenticator.ts [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+}
+function fromRole(role: Partial<Record<"*" | Realm, Key.Permissions.Role>>): Key.Permissions {
+	return Object.fromEntries(
+		(Object.entries(role) as [Realm, Key.Permissions.Role][]).map(([realm, role]) => [
+			realm,
+			Key.Permissions.role[role],
+		])
+	) as Partial<Record<"*" | Realm, Key.Permissions.Realm>>
+}
+function findRealmPermissions(permissions: Key.Permissions, starPermissions: Key.Permissions): Key.Permissions {
```
Comment (marcushanses) [inline Authenticator.ts:52]: I don't understand this function.
You put in permissions and starpermissions?
What's the difference between permissions and starPermissions?
Then you make an object with a random key from permissions which will equal the value of starPeemissions.*.
Please explain.

**Review by marcushanses — CHANGES_REQUESTED** [review]

File: Authenticator.spec.ts [hunk 11]
```diff
@@ -0,0 +1,154 @@
+import { flagly } from "flagly"
+import { userwidgets } from "@userwidgets/model"
+import { pax2pay } from "./index"
+
+describe("authenticator", () => {
+	const privateKey =
```
[the remark this answers is not in the document]
Comment (AntonCroner) [reply Authenticator.spec.ts:6]: done

File: Authenticator.spec.ts (hunk 2 above)
Comment (AntonCroner) [reply Authenticator.spec.ts:23]: done

File: Authenticator.spec.ts (hunk 4 above)
Comment (AntonCroner) [reply Authenticator.spec.ts:34]: done

File: Authenticator.spec.ts (hunk 5 above)
Comment (AntonCroner) [reply Authenticator.spec.ts:20]: done

**Pull request closed without merging**